# Table of Contents

### 1. Import Librairies
### 2. Import Data Sets
### 3. Cleaning the data set
#### Descriptive Statistics
### 4. Export Data

# 1. Import librairies

In [2]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 2. Import Data Sets

In [3]:
#Creating path
path = r'C:\Users\manev\Documents\Agentür für Arbeit\CareerFoundry\Data Immersion\Achievement 4\Instacart Basket Analysis'

In [4]:
#Create data frame for the 'products' data set
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col=False)

In [5]:
#Create data frame for the 'orders_wrangled' data set
df_ords_wrangled = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 3. Cleaning the data set

In [6]:
#Find missing values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [7]:
#See all missing values in the product_name column
df_nan= df_prods[df_prods['product_name'].isnull()== True]

In [8]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [9]:
df_prods.shape

(49693, 5)

In [10]:
#Filtering out missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [11]:
df_prods_clean.shape

(49677, 5)

In [12]:
#Lookingfor duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [13]:
 df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [14]:
#Checking number of rows before removing duplicates
df_prods_clean.shape

(49677, 5)

In [15]:
#Creating a new dataframe without duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [16]:
df_prods_clean_no_dups.shape


(49672, 5)

In [17]:
#Export the cleaned version of df_prods_clean_no_dups
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'),index=False)

### Descriptive Statistics

In [18]:
#Examining the descriptive statistics of the df_ords_wrangled data set
df_ords_wrangled.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [19]:
#Examining the descriptive statistics of the df_ords_wrangled data set and show the results without using scientific notation
df_ords_wrangled.describe().apply(lambda x: x.apply('{0:.5f}'.format))

order_id        user_id   order_number orders_day_of_week  \
count  3421083.00000  3421083.00000  3421083.00000      3421083.00000   
mean   1710542.00000   102978.20806       17.15486            2.77622   
std     987581.73982    59533.71779       17.73316            2.04683   
min          1.00000        1.00000        1.00000            0.00000   
25%     855271.50000    51394.00000        5.00000            1.00000   
50%    1710542.00000   102689.00000       11.00000            3.00000   
75%    2565812.50000   154385.00000       23.00000            5.00000   
max    3421083.00000   206209.00000      100.00000            6.00000   

      order_hour_of_day days_since_prior_order  
count     3421083.00000          3214874.00000  
mean           13.45202               11.11484  
std             4.22609                9.20674  
min             0.00000                0.00000  
25%            10.00000                4.00000  
50%            13.00000                7.00000  
75%            16.00000               15.00000  
max            23.00000               30.00000

The count of observations in the "days_since_prior_order" column is lower than in the other columns. This suggests that there is some missing values in that column.

In [20]:
#Looking for mixed data in df_ords_wrangled
for col in df_ords_wrangled.columns.tolist():
  weird = (df_ords_wrangled[[col]].map(type) != df_ords_wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords_wrangled[weird]) > 0:
    print (col)

There is no mixed-data in this data frame.

In [21]:
#Running a check for missing values in df_ords_wrangled
df_ords_wrangled.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in the "days_since_prior_order" column. We noticed that the maximum of the "user_id" column is also 206209. We can then assume that the missing values correspond to the first order of each customer (which, understandably, doesn't have a value for "days_since_prior_order", as there is no prior order). 
The missing values are therefore justified.
We could have marked them as "first order", but this would lead to the column having mixed-values. We will therefore leave them as they are.

In [22]:
#Adding a new boolean column to indicate whether an order has been placed by a new customer, thus hustifying the absence of value in the "day_since_prior_order" column
df_ords_wrangled['new_customer'] = df_ords_wrangled['days_since_prior_order'].isnull() == True

In [23]:
df_ords_wrangled

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  new_customer  
0                        8                     NaN          True  
1                        7                    15.0         False  
2                       12                    21.0         False  
3                        7                    29.0         False  
4                       15                    28.0         False  
...                    ...                     ...           ...  
3421078                 18                    29.0         False  
3421079                 10                    30.0         False  
3421080                 12                    18.0         False  
3421081                 12                     7.0         False  
3421082                 14                    30.0         False  

[3421083 rows x 7 columns]

In [24]:
#Looking for duplicates
df_dups = df_ords_wrangled[df_ords_wrangled.duplicated()]

In [25]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, new_customer]
Index: []

No duplicate were found in df_ords_wrangled.

In [26]:
#Check the shape of the rds_wrangled data frame
df_ords_wrangled.shape

(3421083, 7)

# 4. Export Data

In [27]:
#Export the cleaned version of df_ords_wrangled
df_ords_wrangled.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'),index=False)